Fix memory problems in `findMarginalPosteriorDensity`

paciorek opened the pull request:
This fixes two memory/access problems related to `findMarginalPosteriorDensity`.

1. In `findMarginalPosteriorDensity`, we were not respecting that `Atransform_i` is d-1 by d-1 when filling it.
2. In `optimize` in Laplace, we were passing a `d-1` length value into `paramsTransform$inverseTransform`.

For $d=3$, these caused crashes with error messages related to memory use.

Strangely (and probably why not detected until now), for $d=2$, things ran fine.  I am confused why this would be the case, but perhaps it's not important to understand why.

@paul-vdb would be helpful if you'd give this a second eye.

paciorek committed:
Fix handling of includeParams in sampleLatentNodes.

paciorek committed:
Tweak nested approx message.

paciorek committed:
Fix bug in sampling latent nodes with multiple cond ind sets.

paciorek committed:
Add some testing files for nested approx.

paciorek committed:
Fix memory access issues with findMarginalPosteriorDensity.

paul-vdb commented:
@paciorek What is the example where `pTransformed = d-1`? I assume this is when dirichlet or wishart priors are used? I'm surprised if just a simple example of `d=3` was causing an error if `npar` and `pTransform_length` are the same. Will look through this soon.

paciorek commented:
Even in the 1:1 case, when we use `findMarginalPosteriorDensity`, we are maximizing in a $d-1$ dimensional space.

paciorek commented:
Also a note about this PR - I was sloppy and a few other changes not related to the bugs crept in. 

Most importantly, I fixed a separate bug in which we were not passing the initial `nQuadOuter` value set when building the approximation into `calcHyperGrid`, which prevented the user from actually modifying the outer grid. 

I also modified some of the output and messaging syntax and fixed some related stuff in the `runNestedApprox` functionality.

paul-vdb commented:
@paciorek I get it now. This is for when we are doing AGHQ to get the marginals so doing the gradient on d-1 values, holding one fixed, not a general issue with Laplace, except that Laplace didn't have any warnings.

paciorek commented:
@paul-vdb I'm planning to merge this in unless you have further thoughts.

paul-vdb commented:
@paciorek Can we talk about this one? I think I specifically wrote the `keepOneFixed_` parts of Laplace for exactly this reason. I want to double check the details here before we merge to make sure if you have a better solution, we can clean up unneeded code in Laplace etc. I'll email times to chat.

I should also say that is why it was working for d=2 case but not larger. So if you have resolved that and made a better solution, we should remove mine as well.

paciorek commented:
Paul and I got on the same page on Zoom. All set to merge, even though we don't know why `d=2` case was ok.
